packages/kit/src/runtime/client/client.js

```diff
@@ -1220,19 +1235,23 @@ export function create_client({ target, base, trailing_slash }) {
 			addEventListener('beforeunload', (e) => {
 				let should_block = false;
 
-				/** @type {import('types').Navigation & { cancel: () => void }} */
-				const navigation = {
-					from: add_url_properties('from', {
-						params: current.params,
-						routeId: current.route?.id ?? null,
-						url: current.url
-					}),
-					to: null,
-					type: 'unload',
-					cancel: () => (should_block = true)
-				};
+				if (get(stores.navigating) === null) {
+					// If we're navigating, beforeNavigate was already called. If we end up in here during navigation,
+					// it's due to a 404 or another error, for which we don't want to call the hook again.
+					/** @type {import('types').Navigation & { cancel: () => void }} */
```

Comment:
_Is_ this the behaviour we want? I can't decide. If you redirect during a navigation, `beforeNavigate` gets called for the initial navigation then again for each redirect (with `type === 'goto'`, which is arguably incorrect), and this would then be inconsistent with that. You could definitely make the case that we should only skip `beforeNavigate` for external link clicks